elastic/kibana
# [Synthetics] Accept project monitors with `monitor.url` of type `string` that contains commas

## Summary

Resolves #186093.

Result of an [SDH](https://github.com/elastic/sdh-synthetics/issues/97) that lead to a desire to refine Kibana's acceptance criteria for lightweight project monitors.

Heartbeat accepts `monitor.url` values of type `string` as a comma-delimited list, like:

`https://elastic.co,https://amazon.com`

This is not supported for Synthetics, the design goals of the product demand a 1:1 ratio relationship between URL<->Monitor.

Thus the code attempts to handle this parsing. It has the unintended side effect of rejecting otherwise valid URLs that contain commas. Rather than creating even more sophisticated parsing for this case, this patch would remove all parsing of string fields, and simply convert them to a list with one entry, which is what we then store in the monitor's definition.

_NOTE:_ this also adds a call to `new URL` to attempt to parse the URL value. This wasn't present before, but the Agent team made it clear that we should do best-effort checks to ensure that an invalid URL doesn't actually make it to the point of executing on the service. I've added a fourth test case to the testing instructions below.

## Testing

There are several cases we must cover to ensure we aren't breaking anything with this change. We must initiate these checks against both public and private locations, as the backend behavior of the two differs.

- Create a monitor in a list containing a comma. Example:
```yaml
heartbeat.monitors:
- type: http
  name: Todos Lightweight
  id: todos-lightweight
  enabled: true
  urls:
    - "https://elastic.github.io/synthetics-demo/?demo=123,345"
  schedule: '@every 10s'
  timeout: 16s

```

- Create a monitor with the URL defined as a string, containing a comma:
```yaml
heartbeat.monitors:
- type: http
  name: Todos Lightweight
  id: todos-lightweight
  enabled: true
  urls: "https://elastic.github.io/synthetics-demo/?demo=123,345"
  schedule: '@every 10s'
  timeout: 16s
```

- Create a monitor with a URL defined as a string, containing a comma-delimited list of valid URLs:
```yaml
heartbeat.monitors:
- type: http
  name: Trollface two montiors
  id: todos-lightweight-multiple
  enabled: true
  urls: "https://elastic.github.io/synthetics-demo/,https://elastic.co"
  schedule: '@every 10s'
  timeout: 16s
  alert.status.enabled: true
```
- Create a monitor with a URL defined as a string, containing an invalid URL:
```yaml
- type: http
  name: Unhandled text
  id: unhandled-text
  enabled: true
  urls: "unhandledtext"
  schedule: '@every 10s'
  timeout: 16s
  alert.status.enabled: true
```
